_posts/2020-12-21--kinesis-streams-part-2.md

```diff
@@ -0,0 +1,591 @@
+
+As we have learned in [Part 1](https://dev.solita.fi/2020/05/28/kinesis-streams-part-1.html), that data in a Kinesis stream expires after a minimum of 24 hours. Let’s see what happens in such a stream if a batch contains a “bad” record.
+
+In this unfortunate scenario, by default, Lambda will be retrying the “bad” batch unsuccessfully for 24 hours, causing an extensive number of fruitless Lambda invocations. What if you have configured your stream to have a longer retention period. How about 7 days? Or even, one year? Well, in that case, you got yourself a year's worth of some unnecessary Lambda invocations.
+
+Moreover, those fruitless retries will likely cause multiple reprocessing of some of the records. You see, from the perspective of the _ESM_, either the entire batch succeeds or it fails. When a Lambda function throws an error because of that one “bad” record, it causes the **entire batch** to fail and to be retried all over again, even if some of the records in that batch were already successfully processed earlier, before the error occurred.
+
+
+![retries](/img/kinesis/retries.png){: .img.centered }
+
+
+Finally, 24 hours later, the data expires and is removed from the stream, partly unprocessed. When this happens, you end up losing not only that one “bad” record that was causing the problem but the **entire batch** of your valuable data.
+
+But bad things don’t stop there, quite the contrary.
+
+The thing is, while Lambda keeps retrying that one unfortunate batch from the shard, **no other batches are being read from that shard**. This, of course, is a reasonable thing to do while otherwise Kinesis and Lambda would not be able to guarantee ordering of records within a shard. However, it also means that because of just one “bad” record, your entire shard is now stuck, while the rest of the shards in your stream go on with their lives. That’s why this “bad” record is often referred to as a poison pill.
+
+In our scenario, in 24 hours the poison pill finally leaves the shard, together with the rest of the records that were unlucky enough ending up in the same batch. But at that moment when Lambda can finally start taking in new batches, the shard is potentially filled with records that were written to the stream around the same time. This, in turn, means that they **expire around the same time** as the already discarded batch. So, u may end up in a situation, where your Lambda function just doesn’t have enough time to catch up with the shard and your records will keep on falling off the edge of the stream, so to speak (remember our overflowing sink analogy?).
+
+
+![thisisfine](/img/kinesis/this_is_fine.jpg){: .img.centered }
+
+
+So, even though we started off with just one “bad" record which could not be processed, we ended up losing a lot of valid and valuable data. All this because we didn’t have **proper error handling**.
+
+The good news is, there are several ways to mitigate this kind of situations.
```

Comment:
"The good news is that there are several ways to mitigate these kinds of situations."